fedimint-api/src/config.rs

```diff
@@ -19,41 +21,187 @@ use tbs::hash::hash_bytes_to_curve;
 use tbs::poly::Poly;
 use tbs::serde_impl;
 use tbs::Scalar;
+use url::Url;
 
 use crate::cancellable::Cancellable;
 use crate::net::peers::AnyPeerConnections;
-use crate::task::TaskGroup;
-use crate::PeerId;
-
-/// Part of a config that needs to be generated to bootstrap a new federation.
-#[async_trait(? Send)]
-pub trait GenerateConfig: Sized {
-    type Params: ?Sized;
-    type ClientConfig;
-    type ConfigMessage;
-    type ConfigError;
-
-    /// Function that generates the config of all peers locally. This is only meant to be used for
-    /// testing as the generating machine would be a single point of failure/compromise.
-    fn trusted_dealer_gen(
-        peers: &[PeerId],
-        params: &Self::Params,
-        rng: impl RngCore + CryptoRng,
-    ) -> (BTreeMap<PeerId, Self>, Self::ClientConfig);
-
-    fn to_client_config(&self) -> Self::ClientConfig;
-
-    /// Asserts that the public keys in the config are and panics otherwise (no way to recover)
-    fn validate_config(&self, identity: &PeerId);
-
-    async fn distributed_gen(
-        connections: &mut AnyPeerConnections<Self::ConfigMessage>,
-        our_id: &PeerId,
-        peers: &[PeerId],
-        params: &Self::Params,
-        rng: impl RngCore + CryptoRng,
-        task_group: &mut TaskGroup,
-    ) -> Result<Cancellable<(Self, Self::ClientConfig)>, Self::ConfigError>;
+use crate::Amount;
+use crate::{dyn_newtype_define, PeerId};
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct Node {
+    pub url: Url,
+    pub name: String,
+}
+
+/// Total client config
+///
+/// This includes global settings and client-side module configs.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ClientConfig {
+    pub federation_name: String,
+    pub nodes: Vec<Node>,
+    pub modules: BTreeMap<String, ClientModuleConfig>,
```

Comment:
How come modules are referenced by string vs. a u64 id or something?

Reply:
Internally and for conesnsus-like-necoindg modules are going to be using `u16`, but for config reasons that might not be very convenient. Some kind of mapping might be necessary or something.

We will have to figure it out later.